src/collection.jl

```diff
@@ -115,30 +194,252 @@ end
+            if path isa Nothing
+                path = dirname(location)
+            end
+            save_path = generate_filename(location, path, save_prefix, save_suffix, save_delim, ext)
+            writefits(save_path, processed_image)
+        end
+        processed_image
+    end
+
+    return processed_images
+end
+
+
+"""
+    filenames(f,
+              collection;
+              path = nothing,
+              save_prefix = nothing,
+              save_suffix = nothing,
+              save = any(!isnothing, (save_prefix, path, save_suffix)),
+              save_delim = "_",
+              ext = r"fits(\\.tar\\.gz)?"i,
+              kwargs...)
+
+Iterates over the file paths of the collection applying function `f` at each step.
+
+It returns an array of output values of function `f` applied on file paths.
+The output from `f` can be saved using the appropriate keyword arguments. The `save_prefix` argument will add a prefix to each filename delimited by `save_delim`. `save_suffix` will add a suffix prior to the extension, which can be manually provided via `ext`, similar to [`fitscollection`](@ref). Files will be saved in the directory they are stored unless `path` is given. Finally, `save` will default to `true` if any of the previous arguments are set, but can be manually overridden (useful for testing). Files will be saved using [`CCDReduction.writefits`](@ref).
+
+# Examples
+```julia
+collection = fitscollection("~/data/tekdata")
+data = map(filenames(collection)) do path
+    fh = FITS(path)
+    data = getdata(fh[1]) # assuming all 1-hdu are ImageHDUs
+    close(fh)
+    data
+end
+```
+The above generates `data` which consists of image arrays corresponding to 1st hdu of FITS file paths present in `collection`.
+For saving the `data` simultaneously with the operations performed
+```julia
+data = map(filenames(collection; path = "~/data/tekdata", save_prefix = "retrieved_from_filename")) do img
+    fh = FITS(path)
+    data = getdata(fh[1]) # assuming all 1-hdu are ImageHDUs
+    close(fh)
+    data
+end
+```
+The retrieved data is saved as `retrieved_from_filename_(original_name)` (FITS files) at `path = "~/data/tekdata"` as specified by the user.
+
+Mapping version of `filenames` function is interfaced on iterative version of `filenames`, any valid parameter can be passed into iterative version as `kwargs`.
+"""
+function filenames(f,
+                   collection;
+                   path = nothing,
+                   save_prefix = nothing,
+                   save_suffix = nothing,
+                   save = any(!isnothing, (save_prefix, path, save_suffix)),
+                   save_delim = "_",
+                   ext = r"fits(\.tar\.gz)?"i,
+                   kwargs...)
+    path_iterator = filenames(collection; kwargs...)
+    locations = collection.path
+
+    processed_images = map(zip(locations, path_iterator)) do (location, output)
+        processed_image = f(output)
+        if save
```

Comment:
I've just realized we should throw a check here, too (for all 3 versions)
`if save && !isnothing(processed_image)`

Reply:
Shouldn't this be:
`if save && (processed_image isa Array || processed_image isa CCDData)`?
Because we can only save `Arrays` and `CCDData`s.

Reply:
I don't think we should use an if-statement to enforce type checking. We should let multiple dispatch handle that. The reasoning behind checking for `nothing` is functions *that don't have a return value*.

Thinking back on it, we probably shouldn't do this, since it would be weird to try saving something and failing silently- the user wouldn't see any error or warning that the data isn't being saved. If an error gets thrown (a method error for writefits) that'll fail loudly.